rfd/0035-aws-node-join.md

```diff
@@ -0,0 +1,261 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 35 - Simplified Node Joining for AWS
+
+## What
+
+Teleport nodes running on EC2 instances will be able to join a teleport cluster
+without the need to share any secret token with the node.
+
+Instead, the node will provide either a signed
+[EC2 Instance Identity Document](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html)
+or a signed
+[sts:GetCallerIdentity](https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html)
+request that will be used to confirm the AWS account that the EC2 instance is
+running in. With a configured AWS join token on the auth server, the node
+will be allowed to join the cluster based on its AWS account.
+
+## Why
+
+This will make provisioning nodes on EC2 simpler and arguably more secure.
+There is no need to configure static tokens, or to manage the provisioning and
+deployment of dynamic tokens.
+
+## Details
+
+We are currently exploring two ways to implement this, both inspired by [similar
+designs](https://www.vaultproject.io/docs/auth/aws) used by Vault.
+
+The first is the "EC2 Method", which uses signed EC2 Instance Identity Documents
+which EC2 instances can fetch from the AWS
+[IMDSv2](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html).
+
+The second is the "IAM Method", in which the EC2 instance will create and sign
+an HTTPS API request to Amazon's public `sts:GetCallerIdentity` endpoint and
+pass it to the auth server.
+
+### EC2 Method
+
+In place of a join token, nodes will present an EC2 IID (Instance Identity
+Document) to the auth server, with a corresponding PKCS7 signature.
+
+This will be sent along with the other existing parameters in a
+[RegisterUsingTokenRequest](
+https://github.com/gravitational/teleport/blob/d4247cb150d720be97521347b74bf9c526ae869f/lib/auth/auth.go#L1538-L1563).
+
+The auth server will then:
+1. Use the AWS DSA public certificate to check that the PKCS7 signature for the
```

Comment:
Will Teleport trust the `region` specified in the IID and use that to pick the specific AWS DSA public certificate? Or will this be region-locked via the teleport config? (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/verify-pkcs7.html)

Reply:
I'm mainly worried about the potential security risk of having the attacker "pick" the certificate. Such risks would be mainly related to the implementation of the code we'll see later (e.g. path injections/traversal when retrieving the cert from disk or network). On a different note, an operator may need to restrict this (e.g. Govcloud instances, existing IAM policies that deny access to any actions outside the Regions specified for defense in depth), so I would consider the pros/cons of optionally restrict this.

Reply:
Good point, I've added a configuration option to restrict the regions from which a node can join for the EC2 method (we won't have this info for the IAM method as it isn't included in the `sts:GetCallerIdentity` response)